File: 25.1.zig
const std = @import("std");
const data = @embedFile("input25.txt");

const wt = u140;
const ht = u137;

const width = 140;
const height = 137;

pub fn main() !void {
    var it = std.mem.tokenize(u8, data, "\r\n");

    var rows = [_]u140 {0} ** 137;
    var cols = [_]u137 {0} ** 140;
    _ = cols;

    var line_n: usize = 0;
    while (it.next()) |line| : (line_n += 1){
        for (line) |c, x| {
            if (c == '>') {
                rows[line_n] |= @as(wt, 1) <<% @intCast(u8, x);
            }
            else if (c == 'v') {
                cols[x] |= @as(ht, 1) <<% @intCast(u8, height);
            }
        }
    }

    var moved = true;
    var i: usize = 0;
    while (moved) : (i += 1) {
        for (rows) |row, y| {
            var blocked: wt = 0;
            for (cols) |col| {
                blocked |= (col >> y) & 1;
            }
            _ = row;
        }
    }

    std.debug.print("{d}\n", .{ i / 2 });
}

File: 25.1.cpp
// reading a text file
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <bit>
#include <bitset>
using namespace std;

#define width 139
#define height 137

template <int size>
inline bitset<size> rotLeft(bitset<size> bits) {
    return bits << 1 | bits >> (size - 1);
}

template <int size>
inline bitset<size> rotRight(bitset<size> bits) {
    return bits >> 1 | bits << (size - 1);
}

template <int w, int h>
bool update(vector<bitset<w>> &rows, const vector<bitset<h>> &cols) {
    bool moved = false;
    for (int row = 0; row < rows.size(); row++) {
        bitset<w> blocked;
        for (int col = 0; col < cols.size(); col++) {
            blocked[col] = cols[col][row];
        }

        blocked = (blocked | rows[row]);
        const auto a = rotLeft(rows[row]);
        const auto b = a & (~blocked);
        const auto c = rotRight(b);
        bitset<w> result = rows[row] ^ b ^ c;

        if (result != rows[row]) {
            moved = true;
        }
        rows[row] = result;
    }

    return moved;
}

template <int w, int h>
void print(const vector<bitset<w>> &rows, const vector<bitset<h>> &cols) {
    for (int row = 0; row < height; row++) {
        for (int col = 0; col < width; col++) {
            if (rows[row][col]) {
                cout << '>';
            } else if (cols[col][row]) {
                cout << 'v';
            } else {
                cout << '.';
            }
        }
        cout << '\n';
    }

    cout << '\n' << endl;
}

int main () {
    string line;
    ifstream myfile ("input25.txt");

    vector<bitset<width>> rows(height);
    vector<bitset<height>> cols(width);

    int line_n = 0;
    while (getline (myfile, line)) {
        for (int i = 0; i < line.size(); i++) {
            const char c = line[i];

            if (c == '>') rows[line_n][i] = true;
            if (c == 'v') cols[i][line_n] = true;
        }

        line_n++;
    }

    bool moved = true;
    int runs = 0;
    while (moved) {
        moved = false;
        moved |= update(rows, cols);
        moved |= update(cols, rows);

        runs++;
        if (false) {
            cout << "After " << runs << " steps\n";
            print(rows, cols);
        }
    }

    print(rows, cols);
    cout << runs << endl;

    return 0;
}